utilities/cloudharness_utilities/codefresh.py

```diff
@@ -45,48 +55,71 @@ def codefresh_build_step_from_base_path(base_path, build_step, fixed_context=Non
                 if include and not any(inc in dockerfile_path for inc in include):
                     continue
-                build = codefresh_app_build_spec(
-                    app_name=app_name,
-                    app_context_path=os.path.relpath(fixed_context, '.') if fixed_context else app_relative_to_root,
-                    dockerfile_path=os.path.join(os.path.relpath(dockerfile_path, fixed_context) if fixed_context else '',
-                                                 "Dockerfile"))
-                codefresh['steps'][build_step]['steps'][app_name] = build
-
-        codefresh_build_step_from_base_path(os.path.join(root_path, BASE_IMAGES_PATH), BUILD_STEP_BASE,
+                if CF_BUILD_STEP_BASE in codefresh['steps']:
+                    build = codefresh_app_build_spec(
+                        app_name=app_name,
+                        app_context_path=os.path.relpath(fixed_context, '.') if fixed_context else app_relative_to_root,
+                        dockerfile_path=os.path.join(
+                            os.path.relpath(dockerfile_path, fixed_context) if fixed_context else '',
+                            "Dockerfile"))
+                    codefresh['steps'][build_step]['steps'][app_name] = build
+                if CF_STEP_PUBLISH in codefresh['steps']:
+                    codefresh['steps'][CF_STEP_PUBLISH]['steps'][app_name] = codefresh_app_publish_spec(
+                        app_name=app_name)
+
+        codefresh_build_step_from_base_path(os.path.join(root_path, BASE_IMAGES_PATH), CF_BUILD_STEP_BASE,
                                             fixed_context=root_path)
-        codefresh_build_step_from_base_path(os.path.join(root_path, STATIC_IMAGES_PATH), BUILD_STEP_STATIC)
-        codefresh_build_step_from_base_path(os.path.join(root_path, APPS_PATH), BUILD_STEP_PARALLEL)
+        codefresh_build_step_from_base_path(os.path.join(root_path, STATIC_IMAGES_PATH), CF_BUILD_STEP_STATIC)
+        codefresh_build_step_from_base_path(os.path.join(root_path, APPS_PATH), CF_BUILD_STEP_PARALLEL)
 
     codefresh['steps'] = {k: step for k, step in codefresh['steps'].items() if
                           'type' not in step or step['type'] != 'parallel' or (
                               step['steps'] if 'steps' in step else [])}
 
-    codefresh_abs_path = os.path.join(os.getcwd(), DEPLOYMENT_PATH, codefresh_path)
+    codefresh_abs_path = os.path.join(os.getcwd(), DEPLOYMENT_PATH, out_filename)
     codefresh_dir = os.path.dirname(codefresh_abs_path)
     if not os.path.exists(codefresh_dir):
         os.makedirs(codefresh_dir)
     with open(codefresh_abs_path, 'w') as f:
         yaml.dump(codefresh, f)
 
 
-def codefresh_build_spec(**kwargs):
+def codefresh_template_spec(template_path, **kwargs):
     """
     Create Codefresh build specification
     :return:
     """
 
-    build = get_template(CODEFRESH_BUILD_PATH)
+    build = get_template(template_path)
```

Comment:
I would do here ```return get_template(template_path).update(kwargs)```

Reply:
update is a mutating method and returns null.

Reply:
update is a mutating method and returns None

Reply:
hahah yeah you're right (shame)